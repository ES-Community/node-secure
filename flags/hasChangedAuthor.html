<p>
    The package original author/owner has been updated. This may indicate
    ownership transfer !
</p>
<blockquote>
    <p>
        ⚠️ this flag is not yet as revelant as we want because sometimes we fail
        to retrieve the real package owner.
    </p>
</blockquote>
