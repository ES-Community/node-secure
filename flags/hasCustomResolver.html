<p>
    The package has custom dependencies resolver such as <code>+git</code> or
    <code>+ssh</code> or a local file with <code>file:</code>. Note that pacote
    doesn't support <code>ssh</code> so there is no support in nsecure for this
    kind of resolver.
</p>
<p>
    Documentation:
    <a href="https://docs.npmjs.com/cli/install" rel="nofollow">npm-install</a>
</p>
