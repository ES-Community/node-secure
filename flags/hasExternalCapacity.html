<p>
    The package use a Node.js core package that allow to access the network.
    These core package are:
</p>
<ul>
    <li>http</li>
    <li>https</li>
    <li>net</li>
    <li>http2</li>
    <li>dgram</li>
</ul>
<p>
    This flag only work if the AST analysis as successfully retrieved all
    dependencies as expected.
</p>
