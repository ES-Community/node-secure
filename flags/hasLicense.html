<br />
<p>
    This flag mean that we have not detected any licenses in the npm Tarball (or
    something went wrong in the detection) For detecting licenses we are reading
    the <strong>package.json</strong> and searching for local files that contain
    the word "license".
</p>
<p>
    For more information on how license must be described in the package.json,
    please check the
    <a href="https://docs.npmjs.com/files/package.json#license" rel="nofollow"
        >npm documentation</a
    >.
</p>
<blockquote>
    <p>
        ⚠️ we are working to stabilize this flag !
    </p>
</blockquote>
