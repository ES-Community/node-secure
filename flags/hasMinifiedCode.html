<p>
    Has one or many files that has been detected as minified JavaScript code. We
    use a package that will tell us if the code is minified (in case the file as
    a <strong>.min</strong> then we will consider the file minified by default).
</p>
<br />
<p>
    Minified JavaScript code are commonly used by hacker to obfuscate the code
    to avoid being spotted. A good practice is surely to check all the packages
    with the flag.
</p>
<br />
<p>Example of minified code:</p>
<br />
<pre><code>function cleanRange(version){const firstChar=version.charAt(0);if(firstChar==="^"||firstChar==="<"||firstChar===">"||firstChar==="="||firstChar==="~"){return version.slice(version.charAt(1)==="="?2:1)}
   return version}</code></pre>
<blockquote>
    <p>
        ⚠️ sometimes one line file are considered minified (we are working to
        fix this in the future).
    </p>
</blockquote>
