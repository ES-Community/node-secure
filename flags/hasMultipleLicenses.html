<p>
    We have detected different licenses in <strong>package.json</strong> and
    other licenses files (<strong>LICENSE</strong>,
    <strong>LICENSE.MD</strong> etc). This probably means that there is an
    inconsistency in the choice of the license (or a file not updated yet with
    the right license).
</p>
<p>
    This flag has not been created to detect multiple licenses / conformance
    rules. Example: <code>ISC OR GPL-2.0-with-GCC-exception</code>. Under the
    hood we use
    <a href="https://github.com/cutenode/conformance#readme">conformance</a> to
    parse licenses !
</p>
