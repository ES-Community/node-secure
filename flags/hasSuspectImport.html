<br />
<p>
    This mean that AST (Abstract Syntax Tree) analysis of one of the JavaScript
    code of the package has failed because the code was to complicated. For
    example if your package contains a .js file with the following content:
</p>
<br />
<pre><code>const { readFileSync } = require("fs");
const { join } = require("path");
const myLib = require("./lib");</code></pre>
<br />
<p>
    Then the AST analysis will return <code>fs</code>, <code>path</code> and
    <code>./lib</code> as required dependencies. The code will not be considered
    suspicious ! But if we take a malicious code:
</p>
<pre><code>function unhex(r) {
   return Buffer.from(r, "hex").toString();
}

const g = Function("return this")();
const p = g["pro" + "cess"];

const evil = p["mainMod" + "ule"][unhex("72657175697265")];
evil(unhex("68747470")).request</code></pre>
<p>
    This code require the core package http but the AST analysis is not capable
    to get it (not yet üòÅ). So the code will be flagged as "suspect".
</p>
