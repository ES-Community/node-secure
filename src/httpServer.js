/* eslint-disable no-sync */

// Import Node.js Dependencies
import fs from "fs";
import { readFile } from "fs/promises";
import { join } from "path";
import { pipeline } from "stream";

// Import Third-party Dependencies
import send from "@polka/send-type";
import kleur from "kleur";
import polka from "polka";
import sirv from "sirv";
import open from "open";
import zup from "zup";
import i18n from "@nodesecure/i18n";
import { getFlags, getFlagFile, getManifest } from "@nodesecure/flags";

// CONSTANTS
const kNodeSecureFlags = getFlags();
const kProjectRootDir = join(__dirname, "..");

export async function startHTTPServer(dataFilePath, configPort) {
  fs.accessSync(dataFilePath, fs.constants.R_OK | fs.constants.W_OK);

  const httpServer = polka();
  httpServer.use(sirv(join(kProjectRootDir, "dist"), { dev: true }));

  httpServer.get("/", async(req, res) => {
    try {
      res.writeHead(200, {
        "Content-Type": "text/html"
      });

      const HTMLStr = await readFile(join(kProjectRootDir, "views", "index.html"), "utf-8");
      const templateStr = zup(HTMLStr)({
        lang: i18n.getToken("lang"),
        token: (tokenName) => i18n.getToken(`ui.${tokenName}`)
      });

      res.end(templateStr);
    }
    catch (err) {
      send(res, 500, { error: err.message });
    }
  });

  httpServer.get("/data", (req, res) => {
    res.writeHead(200, { "Content-Type": "application/json" });
    pipeline(fs.createReadStream(dataFilePath), res, (err) => {
      if (err) {
        console.error(err);
      }
    });
  });

  httpServer.get("/flags", (req, res) => send(res, 200, getManifest()));
  httpServer.get("/flags/description/:title", (req, res) => {
    if (req.params.title !== "isDuplicate" && !kNodeSecureFlags.has(req.params.title)) {
      return send(res, 404, { error: "Not Found" });
    }

    res.writeHead(200, { "Content-Type": "text/html" });

    return pipeline(getFlagFile(req.params.title), res, (err) => {
      if (err) {
        console.error(err);
      }
    });
  });

  /* istanbul ignore next */
  httpServer.listen(typeof configPort === "number" ? configPort : 0, () => {
    const link = `http://localhost:${httpServer.server.address().port}`;
    console.log(kleur.magenta().bold(i18n.getToken("cli.http_server_started")), kleur.cyan().bold(link));

    if (typeof configPort === "undefined") {
      open(link);
    }
  });

  return httpServer;
}
