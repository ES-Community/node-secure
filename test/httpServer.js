// Require Node.js Dependencies
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import path from "path";

// Require Third-party Dependencies
import { get } from "httpie";
import zup from "zup";
import * as i18n from "@nodesecure/i18n";
import * as flags from "@nodesecure/flags";

// Require Internal Dependencies
import { startHTTPServer } from "../src/httpServer.js";

// CONSTANTS
const HTTP_PORT = 1337;
const HTTP_URL = new URL(`http://localhost:${HTTP_PORT}`);
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const JSON_PATH = path.join(__dirname, "fixtures", "httpServer.json");
const INDEX_HTML = readFileSync(path.join(__dirname, "..", "views", "index.html"), "utf-8");

// VARS
let httpServer;

beforeAll(async() => {
  httpServer = await startHTTPServer(JSON_PATH, {
    configPort: HTTP_PORT,
    openLink: false
  });
});

afterAll(() => {
  httpServer.server.close();
});

test("'/' should return index.html HTML content", async() => {
  const result = await get(HTTP_URL);

  expect(result.statusCode).toStrictEqual(200);
  expect(result.headers["content-type"]).toStrictEqual("text/html");

  const templateStr = zup(INDEX_HTML)({
    lang: i18n.getToken("lang"),
    token: (tokenName) => i18n.getToken(`ui.${tokenName}`)
  });
  expect(result.data).toStrictEqual(templateStr);
});

test("'/flags' should return the flags list as JSON", async() => {
  const result = await get(new URL("/flags", HTTP_URL));

  expect(result.statusCode).toStrictEqual(200);
  expect(result.headers["content-type"]).toStrictEqual("application/json;charset=utf-8");
  expect(result.data).toMatchSnapshot();
});

test("'/flags/description/isGit' should return the isGit HTML description", async() => {
  const result = await get(new URL("/flags/description/isGit", HTTP_URL));

  expect(result.statusCode).toStrictEqual(200);
  expect(result.headers["content-type"]).toStrictEqual("text/html");

  expect(result.data).toStrictEqual(await flags.eagerFetchFlagFile("isGit"));
});

test("'/flags/description/foobar' should return a 404 error", async() => {
  expect.assertions(2);
  try {
    await get(new URL("/flags/description/foobar", HTTP_URL));
  }
  catch (error) {
    expect(error.statusCode).toStrictEqual(404);
    expect(error.data.error).toStrictEqual("Not Found");
  }
});

test("'/data' should return the fixture payload we expect", async() => {
  const result = await get(new URL("/data", HTTP_URL));

  expect(result.statusCode).toStrictEqual(200);
  expect(result.headers["content-type"]).toStrictEqual("application/json");
  expect(result.data).toMatchSnapshot();
});
