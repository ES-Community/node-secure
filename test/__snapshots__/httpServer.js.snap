// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'/data' should return the fixture payload we expect 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`'/flags' should return the flags list as JSON 1`] = `
Object {
  "⌚️": Object {
    "title": "isOutdated",
    "tooltipDescription": "The current package version is not equal to the package latest version!",
  },
  "☁️": Object {
    "title": "isGit",
    "tooltipDescription": "The package (project) is a git repository",
  },
  "⚔️": Object {
    "title": "hasBannedFile",
    "tooltipDescription": "The project has at least one sensitive file.",
  },
  "⚠": Object {
    "title": "hasWarnings",
    "tooltipDescription": "The AST analysis has detected warnings (suspect import, unsafe regex ..)",
  },
  "⛔️": Object {
    "title": "isDeprecated",
    "tooltipDescription": "The package is deprecated.",
  },
  "🌍": Object {
    "title": "hasExternalCapacity",
    "tooltipDescription": "The package use at least one Node.js dependency capable to communicate outside or to establish a listening server.",
  },
  "🌲": Object {
    "title": "hasIndirectDependencies",
    "tooltipDescription": "The package have indirect dependencies.",
  },
  "🐲": Object {
    "title": "hasNativeCode",
    "tooltipDescription": "The package use native components (package, file, configuration).",
  },
  "👀": Object {
    "title": "hasMissingOrUnusedDependency",
    "tooltipDescription": "A dependency is missing in package.json or a dependency is installed but never used!",
  },
  "👥": Object {
    "title": "hasChangedAuthor",
    "tooltipDescription": "The author has already changed at least one time.",
  },
  "💀": Object {
    "title": "isDead",
    "tooltipDescription": "The dependency has not received update from at least one year!",
  },
  "💎": Object {
    "title": "hasCustomResolver",
    "tooltipDescription": "The package has dependencies that are not packages.",
  },
  "💕": Object {
    "title": "hasManyPublishers",
    "tooltipDescription": "The package has several publishers.",
  },
  "📚": Object {
    "title": "hasMultipleLicenses",
    "tooltipDescription": "The package has multiple licenses in multiple files",
  },
  "📜": Object {
    "title": "hasLicense",
    "tooltipDescription": "The package does not seem to have a license.",
  },
  "📦": Object {
    "title": "hasScript",
    "tooltipDescription": "has \`post\` and/or \`pre\` (un)install npm script",
  },
  "🔬": Object {
    "title": "hasMinifiedCode",
    "tooltipDescription": "The package seems to have files that are minified/uglified.",
  },
  "🚨": Object {
    "title": "vulnerabilities",
    "tooltipDescription": "The package have one or many vulnerabilities.",
  },
}
`;
